### ISSUE TITLE
Howto debug VBCSCompiler

### ISSUE THREAD

**Issue by BartNetJS** [dotnet/roslyn#15284]
Continue issue https://github.com/dotnet/roslyn/issues/15164

I'm able to debug the VBCSCompile, so far so good

But, when I have unittests with fakes, it seems when fakes.exe is launching ,
a new vbcscompiler.exe and csc.exe instance to build the fake dll is also launched, 
but not the one build from the roslyn solution, but the old one

And it is this vbcscompiler and csc instance that is consuming a lot of cpu and never seems to end

I also used the processmonitor from systinternals to monitor vbcscompiler, and when it is consuming a lot of cpu, i see that threads are created and destroyed continuously

see also included print screens
When I start building my project, you see the csc.exe build from the roslyn solution is lauched: 
![csc2](https://cloud.githubusercontent.com/assets/4865355/20352268/ebfecc5c-ac16-11e6-8f15-4fe0aa72a2c4.PNG)

When the fakes are build, you see another instance of the vbcscompiler is launched from the wrong folder:
![vbcscompiler1](https://cloud.githubusercontent.com/assets/4865355/20352266/ebf9b0dc-ac16-11e6-9a3a-2e72e566a663.PNG)

And here you see the threads that are created and destroyed:
![processmointor](https://cloud.githubusercontent.com/assets/4865355/20352269/ec021b8c-ac16-11e6-91ba-cdf1791b44a1.PNG)

Als a csc.exe from the wrong folder is called
![csc1](https://cloud.githubusercontent.com/assets/4865355/20352267/ebfd587c-ac16-11e6-906f-8914a7c06de8.PNG)

**Comment by jaredpar** [comment]
I'm unfamiliar with fakes.  Do we have some documentation on how they use the compiler? 

> And it is this vbcscompiler and csc instance that is consuming a lot of cpu and never seems to end

Can you grab a profile of the compilers here?  That should help us quickly see what is going wrong.

**Comment by BartNetJS** [comment]
Here is a link to fakes: https://msdn.microsoft.com/en-us/library/hh549175.aspx

I've also uploaded 2 profiler files from visual studio.
One for the VBCSCompiler.exe and one for CSC.exe

What you see is that almost all the time is going to call the function:
 Microsoft.CodeAnalysis.CSharp.OverloadResolution.PartitionMembersByContainingType

Here a print screen from CSC.EXE when consuming a lot of CPU:

![image](https://cloud.githubusercontent.com/assets/4865355/20386726/2b6c0b2e-acbd-11e6-890c-fa74f37146d1.png)

And for the VBCSCompiler the same:

![image](https://cloud.githubusercontent.com/assets/4865355/20386861/be6314cc-acbd-11e6-96e1-49b393607d31.png)

I've uploaded the profiler files to https://github.com/BartNetJS/VBCSCompilerIssue

Can I grap somewhere the pdb files for VBCSCompiler and CSC?
Then I can debug and give more insights

**Comment by jaredpar** [comment]
The Roslyn symbols should be available on the public symbol server.  Navigate to Tools -> Options -> Debugging -> Symbols and check that you have the "Microsoft Symbol Server" option checked. 

I dug into this trace a bit and unfortunately the NGEN images seem to be confusing the profiler.  I can't get any accurate names out of it hence I can't quite see why we are looping so hard in that path.   Can you try and grab a trace with PerfView instead? 

> https://www.microsoft.com/en-us/download/details.aspx?id=28567

That will produce a series of ETL files that shouldn't have this issue when digging into.

**Comment by jaredpar** [comment]
Also what version of VS2015 are you using: RTM, update 1, etc ...?

**Comment by BartNetJS** [comment]
Microsoft Visual Studio Enterprise 2015
Version 14.0.25431.01 Update 3

I've uploaded the trace file on my OneDrive:
https://1drv.ms/f/s!AokPog9-J9gvh4c4Ur-3h5m4QkbJSA

**Comment by BartNetJS** [comment]
I can't load the symbol files for VBCSCompiler via the microsoft symbol server, but I was able to decompile via ReSharper

What I found so far is that the compile time for the fakes is taking a lot of time:

![image](https://cloud.githubusercontent.com/assets/4865355/20432458/8e511774-ad9f-11e6-98eb-af8f1c144de0.png)

**Comment by BartNetJS** [comment]
one step in a time :-)
I saw while debugging the VBCSCompiler that the stub methods where compiled.
Because I don't use stubs in the tests, only shims, I disabled the stub generation...
And problem gone
Of course, the question is why the stubs are the problem.

For someone with the same problem:
* open the fakes config file for the fakes dll
* for example mscorlib

  ![image](https://cloud.githubusercontent.com/assets/4865355/20433502/33670ec2-ada4-11e6-9a53-bf6b99fcf112.png)

* and past the StubGeneration snipped

```xml
<Fakes xmlns="http://schemas.microsoft.com/fakes/2011/">
  <Assembly Name="mscorlib" Version="4.0.0.0"/>
  <StubGeneration>
    <Clear />
  </StubGeneration>
</Fakes>
```

**Comment by jaredpar** [comment]
@BartNetJS thanks for the extra info here.

How big is the mscorlib fakes file when stubs are included?  Generally compiling mscorlib is expensive due to the size.  It's possible it's just the size here which is causing the problem.  Or that it's being faked incorrectly.  I'm hoping to look at the trace early next week to dig into it.

**Comment by BartNetJS** [comment]
Hi @jaredpar , what I saw is the stub generation resulted in 1000 and 1000 objects that needs to be compiled, and this over multiple threads.

I saw it especially for the DWHProcessManager.BSL(this is a custom legacy project) fakes.
I'm not sure if it was the same for mscorlib. If it is any help I can check it out

**Comment by jaredpar** [comment]
That screenshot of 700+ members to consider in overload resolution is pretty telling.  I'm sure that's at the heart of the problem.

**Comment by jaredpar** [comment]
This problem should be addressed now in 15.3. We've found and fixed a number of perf issues around overload resolution in the face of a large number of candidate methods. If you still see this behavior then please let us known.
